src/tests/core/blocks/test_create_block.c

```diff
@@ -0,0 +1,208 @@
+  tx1->inputs[0].prev_utxo_output = 0;
+  tx1->outputs[0].amt = 90; 
+  mempool_init();
+  mempool_add(tx1);
+}
+
+static char *test_create_coinbase_tx(){
+  unsigned long test = 10;
+  Transaction *tx = create_coinbase_tx(test);
+  mu_assert(
+    "Coinbase Inputs are not NULL",
+    tx->inputs == NULL
+  );
+  mu_assert(
+    "Coinbase num Inputs is not 0",
+    tx->num_inputs == 0
+  );
+  mu_assert(
+    "Coinbase num Outputs is not 1",
+    tx->num_outputs == 1
+  );
+  mu_assert(
+    "Coinbase output amount is not what is expected",
+    tx->outputs[0].amt == test+BLOCK_REWARD
+  );
+  return NULL;
+}
+
+static char *test_get_txs_from_mempool(){
+  _fill_mempool();
+  Transaction **test_ptr = malloc(sizeof(Transaction**));
+  unsigned int num_tx = get_txs_from_mempool(&test_ptr);
+  mu_assert(
+    "Wrong number of transactions",
+    num_tx == 2
+  );
+  mu_assert(
+    "Coinbase TX num_input doesn't match",
+    test_ptr[0]->num_inputs == 0
+  );
+  mu_assert(
+    "Coinbase TX num_outputs doesn't match",
+    test_ptr[0]->num_outputs == 1
+  );
+  mu_assert(
+    "Coinbase TX output amt doesn't match",
+    test_ptr[0]->outputs[0].amt == 10+BLOCK_REWARD
+  );
+  mu_assert(
+    "TX num_input doesn't match",
+    test_ptr[1]->num_inputs == 1
+  );
+  mu_assert(
+    "TX num_outputs doesn't match",
+    test_ptr[1]->num_outputs == 1
+  );
+  mu_assert(
+    "TX output amt doesn't match",
+    test_ptr[1]->outputs[0].amt == 90
+  );
+  unsigned char *total_hash = hash_all_tx(test_ptr,2);
+  unsigned char *total_hash2 = hash_all_tx(test_ptr,2);
+  mu_assert(
+    "Total hash is inconsistent",
+    memcmp(total_hash, total_hash2, ALL_TX_HASH_LEN) == 0
+  );
+
+  BlockHeader *test_header = create_block_header(test_ptr, 2);
+  mu_assert(
+    "Total hash is not consistent in header creation",
+    memcmp(total_hash, test_header->all_tx, ALL_TX_HASH_LEN) == 0
+  );
+  mu_assert(
+    "Header nonce invalid",
+    test_header->nonce == 0
+  );
+  mu_assert(
+    "Header timestamp invalid",
+    test_header->timestamp > 1647573975
+  );
+
+
+  Block *test_block = create_block_alloc();
+  mu_assert(
+    "Block txs coinbase is bad",
+    test_block->txs[0]->num_inputs !=1
+  );
+  mu_assert(
+    "TX output amt doesn't match",
+    test_block->txs[1]->outputs[0].amt == 90
+  );
+  change_nonce(test_block);
+  mu_assert(
+    "Nonce increment failed",
+    test_block->header.nonce == 1
+  );
+  unsigned char header_hash[BLOCK_HASH_LEN];
+  hash_blockheader(header_hash, &(test_block->header));
+  unsigned char header_hash2[BLOCK_HASH_LEN];
+  hash_blockheader(header_hash2, &(test_block->header));
+  mu_assert(
+    "Header hash is inconsistent",
+    memcmp(header_hash, header_hash2, BLOCK_HASH_LEN) == 0
+  );
+  mu_assert(
+    "Header hash failed",
+    try_header_hash(&(test_block->header)) == 1
+  );
+  mu_assert(
+    "Block hash failed",
+    mine_block(test_block) == 1
+  );
+
+  while(mine_block(test_block) != 0){
```

Comment:
I think this functionality should be in a function.

Reply:
I also think we should make sure to test the full pipeline at some point (make block -> collect tx -> mine). Right now block creation and collect txs are decoupled in the test. (although this is handled by adding more functionality to mine)

Reply:
create_block calls get tx.... but yes agreed

Reply:
Ah you're right. Missed that part